wgengine/tsdns/forwarder.go

```diff
@@ -0,0 +1,308 @@
+	// closed signals all goroutines to stop.
+	closed chan struct{}
+	// wg signals when all goroutines have stopped.
+	wg sync.WaitGroup
+
+	// conns are the UDP connections used for forwarding.
+	// A random one is selected for each request, regardless of the target upstream.
+	conns []*net.UDPConn
+
+	mu sync.Mutex
+	// upstreams are the nameserver addresses that should be used for forwarding.
+	upstreams []net.Addr
+	// txMap maps DNS txids to active forwarding records.
+	txMap map[txid]forwardingRecord
+}
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+func newForwarder(logf logger.Logf, responses chan Packet) *forwarder {
+	return &forwarder{
+		logf:      logger.WithPrefix(logf, "forward: "),
+		responses: responses,
+		queue:     make(chan forwardedPacket, forwardQueueSize),
+		closed:    make(chan struct{}),
+		conns:     make([]*net.UDPConn, connCount),
+		txMap:     make(map[txid]forwardingRecord),
+	}
+}
+
+func (f *forwarder) Start() error {
+	var err error
+
+	for i := range f.conns {
+		f.conns[i], err = net.ListenUDP("udp", nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	f.wg.Add(2*connCount + 1)
+	for idx, conn := range f.conns {
+		go f.send()
+		go f.recv(uint16(idx), conn)
+	}
+	go f.cleanMap()
+
+	return nil
+}
+
+func (f *forwarder) Close() {
+	select {
+	case <-f.closed:
+		return
+	default:
+		// continue
+	}
+	close(f.closed)
+
+	for _, conn := range f.conns {
+		conn.SetDeadline(aLongTimeAgo)
+	}
+
+	f.wg.Wait()
+
+	for _, conn := range f.conns {
+		conn.Close()
+	}
+}
+
+func (f *forwarder) setUpstreams(upstreams []net.Addr) {
+	f.mu.Lock()
+	f.upstreams = upstreams
+	f.mu.Unlock()
+}
+
+func (f *forwarder) send() {
+	defer f.wg.Done()
+
+	var packet forwardedPacket
+	for {
+		select {
+		case <-f.closed:
+			return
+		case packet = <-f.queue:
+			// continue
+		}
+
+		connIdx := rand.Intn(connCount)
+		conn := f.conns[connIdx]
+		_, err := conn.WriteTo(packet.payload, packet.dst)
+		if err != nil {
+			// Do not log errors due to expired deadline.
+			if !errors.Is(err, os.ErrDeadlineExceeded) {
+				f.logf("send: %v", err)
+			}
+			return
+		}
+	}
+}
+
+func (f *forwarder) recv(connIdx uint16, conn *net.UDPConn) {
+	defer f.wg.Done()
+
+	for {
+		out := make([]byte, maxResponseBytes)
+		n, err := conn.Read(out)
+
+		if err != nil {
+			// Do not log errors due to expired deadline.
+			if !errors.Is(err, os.ErrDeadlineExceeded) {
+				f.logf("recv: %v", err)
+			}
+			return
+		}
+
+		if n < headerBytes {
+			f.logf("recv: packet too small (%d bytes)", n)
+		}
+
+		out = out[:n]
+		txid := getTxID(out)
+
+		f.mu.Lock()
+
+		record, found := f.txMap[txid]
+		// At most one nameserver will return a response:
+		// the first one to do so will delete txid from the map.
+		if !found {
+			f.mu.Unlock()
+			continue
+		}
+		delete(f.txMap, txid)
+
+		f.mu.Unlock()
+
+		packet := Packet{
+			Payload: out,
+			Addr:    record.src,
+		}
+		select {
+		case <-f.closed:
+			return
+		case f.responses <- packet:
```

Comment:
Why do responses flow back on a channel? I expected them to get immediately resent to the final destination, but I think I'm missing something about the code structure around the resolver.

Reply:
That would work, I imagine. Currently the responses are ultimately polled by `userspaceEngine` and shoved into the TUN. I guess I could bind a socket to `100.100.100.100` and send them directly.

Reply:
Ah, no that won't work, unless 100.100.100.100 is assigned to a local interface you won't be able to bind a socket. Sending it back through the tun makes sense.